### PR TITLE
8288064: Class initialization locking

### PR DESCRIPTION
**Pull request by coleenp**
Use a native JVM monitor and state for mutual exclusion for class linking and initialization.  See CR for more details.
Tested with tier1-8. tiers 1-4 on Oracle supported platforms and 5-8 on linux-x64-debug.  There isn't any platform dependent code in this change.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.java.net/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8288064](https://bugs.openjdk.org/browse/JDK-8288064): Class initialization locking

### Reviewers
 * [Robbin Ehn](https://openjdk.java.net/census#rehn) (@robehn - **Reviewer**) ⚠️ Review applies to [bf2c7aee](https://git.openjdk.org/jdk/pull/9141/files/bf2c7aeeb1be2ce2432cca8898faa0556ec65817)
 * [Vladimir Ivanov](https://openjdk.java.net/census#vlivanov) (@iwanowww - **Reviewer**) ⚠️ Review applies to [bf2c7aee](https://git.openjdk.org/jdk/pull/9141/files/bf2c7aeeb1be2ce2432cca8898faa0556ec65817)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9141/head:pull/9141` \
`$ git checkout pull/9141`

Update a local copy of the PR: \
`$ git checkout pull/9141` \
`$ git pull https://git.openjdk.org/jdk pull/9141/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9141`

View PR using the GUI difftool: \
`$ git pr show -t 9141`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9141.diff">https://git.openjdk.org/jdk/pull/9141.diff</a>

</details>

### CODE REVIEW

**Comment by coleenp** [comment]
Thanks Robbin.  yes it passes the JCK both locally and in our mach5 test system.

**Review by iwanowww — APPROVED** [review]
Looks good.

**Comment by coleenp** [comment]
Thanks Vladimir and thank you for your help with the CodeCache::flush_dependencies() logic.

File: src/hotspot/share/oops/cpCache.cpp [hunk 1]
```diff
@@ -266,7 +266,7 @@ void ConstantPoolCacheEntry::set_direct_or_vtable_call(Bytecodes::Code invoke_co
     }
     if (invoke_code == Bytecodes::_invokestatic) {
       assert(method->method_holder()->is_initialized() ||
-             method->method_holder()->is_reentrant_initialization(Thread::current()),
+             method->method_holder()->is_reentrant(Thread::current()),
```
Comment (dholmes-ora) [inline src/hotspot/share/oops/cpCache.cpp:269]: Please change this back - `is_reentrant` has no meaning by itself, reentrant with respect to what?

File: src/hotspot/share/oops/instanceKlass.cpp [hunk 2]
```diff
@@ -1177,19 +1172,23 @@ void InstanceKlass::initialize_impl(TRAPS) {
...
 
-void InstanceKlass::set_initialization_state_and_notify(ClassState state, TRAPS) {
-  Handle h_init_lock(THREAD, init_lock());
-  if (h_init_lock() != NULL) {
-    ObjectLocker ol(h_init_lock, THREAD);
+void InstanceKlass::set_initialization_state_and_notify(ClassState state, JavaThread* current) {
+  MonitorLocker ml(current, _init_monitor);
+
+  // Now flush all code that assume the class is not linked.
+  // Set state under the Compile_lock also.
```
Comment (dholmes-ora) [inline src/hotspot/share/oops/instanceKlass.cpp:1179]: Wow! Why do we have to do this now?

File: src/hotspot/share/runtime/vframe.cpp [hunk 3]
```diff
@@ -226,7 +226,7 @@ void javaVFrame::print_lock_info_on(outputStream* st, int frame_count) {
       Klass* k = obj->klass();
       st->print_cr("\t- %s <" INTPTR_FORMAT "> (a %s)", "parking to wait for ", p2i(obj), k->external_name());
     }
-    else if (thread()->osthread()->get_state() == OBJECT_WAIT) {
+    else if (thread()->osthread()->get_state() == CONDVAR_WAIT) {
       // We are waiting on an Object monitor but Object.wait() isn't the
       // top-frame, so we should be waiting on a Class initialization monitor.
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/vframe.cpp:231]: Comment needs updating. I'm glad to see this still works okay with such a simple change.

File: src/hotspot/share/oops/cpCache.cpp (hunk 1 above)
Comment (coleenp) [reply src/hotspot/share/oops/cpCache.cpp:269]: I can have a is_reentrant_initialization() and an is_reentrant_linking() method if that makes it clearer.

File: src/hotspot/share/oops/instanceKlass.cpp (hunk 2 above)
Comment (coleenp) [reply src/hotspot/share/oops/instanceKlass.cpp:1179]: It's not new, it was moved.

File: src/hotspot/share/runtime/vframe.cpp (hunk 3 above)
Comment (coleenp) [reply src/hotspot/share/runtime/vframe.cpp:231]: Oh, yes, it does need updating.

**Comment by coleenp** [comment]
Thanks David for your comments. I pushed an update and retested with jck tests.

File: src/hotspot/share/oops/instanceKlass.hpp [hunk 4]
```diff
@@ -543,7 +543,8 @@ class InstanceKlass: public Klass {
   bool is_not_initialized() const          { return init_state() <  being_initialized; }
   bool is_being_initialized() const        { return init_state() == being_initialized; }
   bool is_in_error_state() const           { return init_state() == initialization_error; }
-  bool is_reentrant(Thread *thread)        { return thread == _init_thread; }
+  bool is_reentrant_initialization(Thread *thread) { return thread == _init_thread; }
+  bool is_reentrant_linking(Thread *thread)        { return thread == _init_thread; }
```
Comment (dholmes-ora) [inline src/hotspot/share/oops/instanceKlass.hpp:547]: Sorry Coleen when I asked for this I thought these queries included the state check, but now I see the state is checked separately. You may as well just have a single simple query `is_init_thread(JavaThread* current)`.
And wherever we set the `_init_thread` we should have an assert that it is NULL (if not already done). Thanks.

Comment (coleenp) [reply src/hotspot/share/oops/instanceKlass.hpp]: Ok, your suggested name change is good and I added the assert we talked about. JCK vm tests pass but I'm rerunning tier1-4 on this change.

**Comment by coleenp** [comment]
Tests in tier1-4 pass with this new assert, which we expected. @dholmes-ora told me offline these were drive-by comments and I've made the suggested changes.  Thanks David.
/integrate

**Pull request closed without merging**